src/processTargets/targets/BaseTarget.ts

```diff
@@ -0,0 +1,183 @@
+import { Range, Selection, TextEditor } from "vscode";
+import { EditNewContext, Position, Target } from "../../typings/target.types";
+import { selectionFromRange } from "../../util/selectionUtils";
+import { getTokenDelimiters } from "../targetUtil/getTokenDelimiters";
+import { maybeAddDelimiter } from "../targetUtil/maybeAddDelimiter";
+
+export function extractCommonParameters(parameters: CommonTargetParameters) {
+  return {
+    editor: parameters.editor,
+    isReversed: parameters.isReversed,
+    contentRange: parameters.contentRange,
+    position: parameters.position,
+    thatTarget: parameters.thatTarget,
+  };
+}
+
+/** Parameters supported by all target classes */
+export interface CommonTargetParameters {
+  readonly editor: TextEditor;
+  readonly isReversed: boolean;
+  readonly contentRange: Range;
+  readonly position?: Position;
+  readonly thatTarget?: Target;
+}
+
+export interface CloneWithParameters {
+  readonly position?: Position;
+  readonly thatTarget?: Target;
+}
+
+export interface BaseTargetParameters extends CommonTargetParameters {
+  readonly delimiter: string;
+}
+
+export default abstract class BaseTarget implements Target {
+  protected readonly state: BaseTargetParameters;
+
+  constructor(parameters: BaseTargetParameters) {
+    this.state = {
+      editor: parameters.editor,
+      isReversed: parameters.isReversed,
+      contentRange: parameters.contentRange,
+      position: parameters.position,
+      delimiter: parameters.delimiter,
+      thatTarget: parameters.thatTarget,
+    };
+  }
+
+  get position() {
+    return this.state.position;
+  }
+  get editor() {
+    return this.state.editor;
+  }
+  get isReversed() {
+    return this.state.isReversed;
+  }
+  get contentRemovalRange() {
```

Comment:
Hmm.  This property name is a bit confusing.  It's not immediately clear how it differs from `removalRange`.  It really feels like it should be private, as it's kind of an implementation detail.

I believe it's only used by `ContinuousRangeTarget`, no? Kind of makes me wonder whether all targets should actually be ranges, with today's targets just being the special case where the range consists of the same start and end scope. Although I believe some of our primitive targets already are ranges today, because eg a LineTarget is really a range of lines, right?  If so, then Targets could just implement a function called `makeRange(active, includeAnchor, includeActive)` which would just return another Target of the same type

Fwiw I'm happy for the target to downgrade itself when you form a range if the argument is not of the same type.  Eg if you say "take line air past token bat", it should downgrade to a WeakTarget.

Reply:
now that everything is lazy, couldn't you just union the content ranges in the continuous range stage and return a new target?

Reply:
fwiw when we fall back to weaktarget, I _think_ I would be ok with discarding the leading / trailing delimiters from the anchor / active and just using token delimiters provided by weaktarget if it simplifies things a bunch.  Eg "chuck line air past token bat" could just convert to a token range from start of line air through end of token bat

Reply:
We still have syntax scope target that needs to transfer their delimiters to a new target and can't be done lazily today. 

This question is a bit too complex I think we need to have a discord about it

Reply:
Yeah let's discuss over discord some time today

For syntax node, that could happen in its makeRange, because it's a lot less complex I think than what's happening here to account for lines

Reply:
In case not clear from our discussion, I'd argue that eg for `SyntaxNodeTarget`, it should always have a `startNode`and `endNode`, and it's `makeContinuousRangeWith` function should just return a new `SyntaxNodeTarget` with start node from start and end node from end

Reply:
Also, your unifyRemovalRanges could _potentially_ just use the new make range as well